feat(scraper): create bank_of_mauritius scraper

mlamchamkee opened the pull request:
Resolves: #75
This PR implements a scraper function for Bank of Mauritius to download and save any number of recent circular pdf files.  It is meant to be merged after [PR#15](https://github.com/sgoldfarb2/scraper-infra/pull/15)  It is currently pointing to the branch [decoupleGetFilePath](https://github.com/sgoldfarb2/scraper-infra/tree/decoupleGetFilePath) for a cleaner diff, but should point to main after PR#15 is merged and decoupleGetFilePath is deleted.

**A couple things to note:**
- We may want to check that the URL is correct.  It was not 100% evident that they are circulars or just publications.
- One challenge was that the each PDF is listed individually on a separate page, which I tackled by first getting the links for the pages and then going to each page to get the PDF link.
- I assumed that the list of circulars is sorted in descending order.

**Code style:**
If that is not consistent with the standards/style for this code base, please let me know and I will happily revert.
- Using Google style docstrings

_Random fact: I grew up on the island of Mauritius, home of the extinct Dodo bird_

nattiechan reviewed (changes requested):
Thank you for the PR!

I like that you are already thinking ahead and want to build some abstractions on top of existing scrapers. We also have some of the abstractions planned as well, and the reason why they are not implemented at this point is because:

1. The implementation of the abstraction is highly dependent on the underlying AWS Lambda structure, which we are still in the process of finalizing.
2. Each scraper module is treated as its own separate entity and will most likely have their own AWS Lambda function. It would be best to keep all information for a particular scraper contained within the scraper file itself.
3. We have many contributors who are new to Python and will rely on other scraper files as examples in terms of how the scrapers should get built. Having less abstractions that they need to navigate can ease the burden of the learning curve for Python and Playwright.
4. There are a number of planned improvements (like [this](https://github.com/sgoldfarb2/scraper/issues/99), for example) that we will need to add to existing scrapers before building out more abstractions.

So I would like to ask that you contain the PR to the original scope of the ticket, which is to build the bank of Mauritius scraper with as much of the existing infrastructure provided as possible.

Any questions please feel free to ping me on Discord. Thank you!

mlamchamkee commented:
> Thank you for the PR!
> 
> I like that you are already thinking ahead and want to build some abstractions on top of existing scrapers. We also have some of the abstractions planned as well, and the reason why they are not implemented at this point is because:
> 
> 1. The implementation of the abstraction is highly dependent on the underlying AWS Lambda structure, which we are still in the process of finalizing.
> 2. Each scraper module is treated as its own separate entity and will most likely have their own AWS Lambda function. It would be best to keep all information for a particular scraper contained within the scraper file itself.
> 3. We have many contributors who are new to Python and will rely on other scraper files as examples in terms of how the scrapers should get built. Having less abstractions that they need to navigate can ease the burden of the learning curve for Python and Playwright.
> 4. There are a number of planned improvements (like [this](https://github.com/sgoldfarb2/scraper/issues/99), for example) that we will need to add to existing scrapers before building out more abstractions.
> 
> So I would like to ask that you contain the PR to the original scope of the ticket, which is to build the bank of Mauritius scraper with as much of the existing infrastructure provided as possible.
> 
> Any questions please feel free to ping me on Discord. Thank you!

Thank you for the review, really appreciate the time and thoughtfulness!  I have implemented the requested changes and updated the PR description to reflect as such.  Thanks!!!

nattiechan commented on scrapers/bank_of_mauritius.py:
```diff
@@ -0,0 +1,81 @@
+        pdf_links = []
+        for link in links:
+            page.goto(f"{base_url}{link}")
+            pdfs = page.locator("table.tblDownloadFile > tbody > tr > td > a").all()
```
Not gonna lie - the way they did this is kind of disgusting.

Part of me wonders if we can identify the PDF icon and go up one parent, but honestly it's just as fragile so I am okay with this

mlamchamkee replied:
I like that suggestion and went ahead to implement it!

nattiechan reviewed (dismissed):
Looks great!

nattiechan commented:
@mlamchamkee - I would prefer not merging this until the PR base changed to `main` if that is alright. That way we can get an additional reviewer to look at your PR

mlamchamkee commented:
> @mlamchamkee - I would prefer not merging this until the PR base changed to `main` if that is alright. That way we can get an additional reviewer to look at your PR

100% alright! That was my intent

qpgdev reviewed (commented):
Pulled and tested on my machine, and it grabs the correct docs.  I like the decoupling of the logic for a more complex scrape like this, and good job implementing the feedback!

@nattiechan, is there a need to implement the validate_and_get_download_link from utils.py for validation here, or do you think the conditional on 63-64 is sufficient for now?

nattiechan commented on scrapers/bank_of_mauritius.py:
```diff
@@ -0,0 +1,81 @@
+        pdf_links = []
+        for link in links:
+            page.goto(f"{base_url}{link}")
+            pdf_icons = page.locator('img[src="/sites/default/files/pdf-icon.png"]').all()
```
Unsure if this will work or not - what if you tried something along the lines of

```suggestion
            pdf_icons = page.locator('img[contains(@src, "pdf-icon.png")]').all()
```

Probably makes the locator a bit more robust against minute changes like how the folders are structured

mlamchamkee replied:
Unfortunately, this didn't work.  Would you happen to have a link to the Playwright docs for the `locator` method?  The [docs I've found](https://playwright.dev/python/docs/locators) are only on locator methods such as `get_by_role()`, etc..
Stack trace:
```
Exception has occurred: Error
Unsupported token "@src" while parsing selector "img[contains(@src, "pdf-icon.png")]"
  File "/Users/mlamchamkee/dev/scraper-infra/scrapers/bank_of_mauritius.py", line 43, in get_pdf_links
    pdf_icons = page.locator('img[contains(@src, "pdf-icon.png")]').all()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mlamchamkee/dev/scraper-infra/scrapers/bank_of_mauritius.py", line 62, in bank_of_mauritius_scraper
    link = get_pdf_links(BASE_URL, CIRCULAR_PATH)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mlamchamkee/dev/scraper-infra/scrapers/bank_of_mauritius.py", line 82, in <module>
    bank_of_mauritius_scraper()
playwright._impl._errors.Error: Unsupported token "@src" while parsing selector "img[contains(@src, "pdf-icon.png")]"
```

qpgdev replied:
Here's a link to the docs for the page.locator() method (between isClosed and mainFrame):
[https://playwright.dev/docs/api/class-page#page-locator](https://playwright.dev/docs/api/class-page#page-locator)

nattiechan replied:
Right - I forgot some syntax specifically for XPath.

Try some of the following:

```
# Best to have either single quotation on the outside and then double on the inside
# or vice versa. Not double quotation for both
page.locator('xpath=//img[contains(@src, "pdf-icon.png")]')
page.locator('//img[contains(@src, "pdf-icon.png")]')
```

nattiechan replied:
https://playwright.dev/docs/locators#locate-by-css-or-xpath

nattiechan commented on scrapers/bank_of_mauritius.py:
```diff
@@ -0,0 +1,81 @@
+        for link in links:
+            page.goto(f"{base_url}{link}")
+            pdf_icons = page.locator('img[src="/sites/default/files/pdf-icon.png"]').all()
+            pdf_links += [
+                icon.locator("..").get_attribute("href")
+                for icon in pdf_icons
+                if icon.locator("..").get_attribute("href").endswith(".pdf")
+            ]
```
I am generally cool with one-liners, and I notice that devs who write more Java have much higher comfort around complex one-liners.

But here we are doing `icon.locator("..").get_attribute("href")` twice - that is a complexity that is arguably harder to maintain.

```suggestion
      for icon in pdf_icons:
          href = icon.locator("..").get_attribute("href")
          # Unsure if the walrus operator would work here
          # If yes, it would be pretty cool lol
          if href.endswith(".pdf"):
              pdf_links.append(href)
```

mlamchamkee replied:
Thanks for the suggestion!  I had defaulted to a list comprehension for speed, but I agree that the clarity outweighs the minor speed gain here.

nattiechan commented:
> is there a need to implement the validate_and_get_download_link from utils.py for validation here, or do you think the conditional on 63-64 is sufficient for now?

@qpgdev you are a reviewer - you have a voice and the power to make suggestions. What do you think? ;)

qpgdev commented:
> > is there a need to implement the validate_and_get_download_link from utils.py for validation here, or do you think the conditional on 63-64 is sufficient for now?
> 
> @qpgdev you are a reviewer - you have a voice and the power to make suggestions. What do you think? ;)

I would implement the validate_and_get_download_link from utils.py for validation here.

qpgdev commented on scrapers/bank_of_mauritius.py:
```diff
@@ -0,0 +1,79 @@
+    session = requests.Session()
+
+    link = get_pdf_links(BASE_URL, CIRCULAR_PATH, 3)[0]
+
```
I would implement the validate_and_get_download_link from utils.py for validation here, but otherwise pretty happy with this!

nattiechan commented on scrapers/bank_of_mauritius.py:
```diff
@@ -0,0 +1,81 @@
+
+    link = get_pdf_links(BASE_URL, CIRCULAR_PATH)[0]
+
+    try:
+        absolute_url = validate_and_get_download_link(BASE_URL, link)
+    except:
+        raise
```
This try/except is redundant here. The reason why others have it here is because they need to close the browser. Since you close it before hand you can just do

```suggestion
      absolute_url = validate_and_get_download_link(BASE_URL, link)
```

mlamchamkee replied:
apologies, silly me!

nattiechan reviewed (changes requested):
One last thing!

qpgdev reviewed (approved):
Great job refactoring!
